044-cluster-wide-volume-usage-quota-management.md

```diff
@@ -0,0 +1,275 @@
+under [KIP-73](https://cwiki.apache.org/confluence/display/KAFKA/KIP-73+Replication+Quotas). Unfortunately replication
+quotas are designed to manage the additional network load of migrating replicas between brokers, which does not address
+client publication leading to out of disk conditions through replication. As replication throttling is configured in terms
+of `TopicPartitions` the quota plug-in would need a mechanism to translate a logDir into a set of TopicPartitions.
+Assuming it could select generate the appropriate throttle configuration this could lead to unpredictable latency spikes
+for producers configured with `acks >= 1` as the replication from partition leader to follower maybe delayed due to the 
+throttle.
+
+Currently, the kafka-quotas-plugin considers the total quantity of storage and how much of that is used (
+see [issue#2](https://github.com/strimzi/kafka-quotas-plugin/issues/2)) when considering whether to apply throttling to
+clients. This is problematic with respect to handling disk failure for Just a Bunch Of Disks (JBOD) deployments [(KIP-112)](https://cwiki.apache.org/confluence/display/KAFKA/KIP-112%3A+Handle+disk+failure+for+JBOD) as the
+broker will take partitions offline when the volume they are stored on runs out of space. Which in the case of unbalanced usage
+between volumes can lead to a volume running out of storage without throttling being applied.
+
+Throttling down message production on all broker nodes will protect the cluster from
+running out of disk due to replication.
+
+## Motivation
+
+Users need better protection against running out of disk on any of their nodes as it can degrade the service
+in unpredictable ways (corrupted segment logs) which could impede the interventions required to recover.
+
+## Proposal
+
+1. Extend the kafka-quotas-plugin so that we can choose to observe the disk usage 
+of all brokers in the cluster.
+2. [Deprecate](#compatibility) the existing local disk observations.
+3. Add new limit types, so that it is explicit what is being limited and better support heterogeneous disks
+   1. throttle if used bytes exceeds threshold on any volume
+   2. throttle if available bytes less than threshold on any volume
+   3. throttle if available ratio less than threshold on any volume
+4. Introduce extension points in the quotas-plugin to support pluggable sources of quotas and throttle factors
+5. Introduce a fallback throttle factor in case we cannot retrieve the cluster state
+
+So every broker will make its own independent throttling decision based on knowledge of the volumes on all active broker nodes.
+The brokers should all operate on a similar view of the cluster state and make a deterministic decision. If a broker detects that
+any volume in the cluster is becoming too full it will throttle production of messages. The kafka quota API isn't rich 
+enough to do anything smarter about only throttling writes to the brokers running out of space, so we fence the whole cluster.
```

Comment:
In Kraft, the idea has one additional merit.  It protects against out of disk space conditions being cause by metadata operations like create topic.  The authorizer API could check for disk space before allowing operations should as createTopic.

Reply:
You would need to figure out how to wrap the Authorizer API to make it work for any authorizer. And you would need to do it without too much of a performance penalty. 

I guess it would also not allow slowing down the producrs but go straight to blocking them? On the other hand, it might allow much batter handling of JBOD storage and only some brokers getting the disks full since it can deal with that on topic per topic basis.

Reply:
> You would need to figure out how to wrap the Authorizer API to make it work for any authorizer. And you would need to do it without too much of a performance penalty.

Yes, the authorizers would need to be chained somehow, so this storage checking impl could be used alongside whatever actual authorizer implementation the customer required,

 > I guess it would also not allow slowing down the producrs but go straight to blocking them? 

Yes - just go straight to blocking them (with an ACL deny) was the idea I was expressing. 

I see no viable way for an authorizer and quota plugin implementations to collaborate (i.e. authorizer somehow sets some state somewhere that the quota plugin would then react to).

Reply:
> In Kraft, the idea has one additional merit. It protects against out of disk space conditions being cause by metadata operations like create topic. The authorizer API could check for disk space before allowing operations should as createTopic.

There are already [throttles on some admin operations](https://cwiki.apache.org/confluence/display/KAFKA/KIP-599%3A+Throttle+Create+Topic%2C+Create+Partition+and+Delete+Topic+Operations). It doesn't prevent disk exhaustion, of course, but frankly assuming you've sized your KRaft controller disks properly and configured snapshots this should just never happen. 

I'd also be wary of throttling all updated to the __cluster_metadata log like this. In the event appends were completely blocked it would prevent you from taking actions (like reconfiguring snapshots) that would help fix the situation. It would also prevent brokers from updating the ISR, or appending [NoOp records](https://cwiki.apache.org/confluence/display/KAFKA/KIP-835%3A+Monitor+KRaft+Controller+Quorum+Health).

Reply:
Pushed up a section adding it to the rejected alternatives to capture this. We rejected mostly on the strength of it being misleading or a surprising use of authorization.